core/src/main/java/org/elasticsearch/action/update/UpdateResponse.java

```diff
@@ -109,4 +117,65 @@ public String toString() {
         builder.append(",shards=").append(getShardInfo());
         return builder.append("]").toString();
     }
+
+
+    /**
+     * ConstructingObjectParser used to parse the {@link UpdateResponse}. We use a ObjectParser here
+     * because most fields are parsed by the parent abstract class {@link DocWriteResponse} and it's
+     * not easy to parse part of the fields in the parent class and other fields in the children class
+     * using the usual streamed parsing method.
+     */
+    private static final ConstructingObjectParser<UpdateResponse, Void> PARSER;
+    static {
+        PARSER = new ConstructingObjectParser<>(UpdateResponse.class.getName(),
+                args -> {
+                    // index uuid and shard id are unknown and can't be parsed back for now.
+                    String index = (String) args[0];
+                    ShardId shardId = new ShardId(new Index(index, IndexMetaData.INDEX_UUID_NA_VALUE), -1);
+                    String type = (String) args[1];
+                    String id = (String) args[2];
+                    long version = (long) args[3];
+                    ShardInfo shardInfo = (ShardInfo) args[5];
+                    Long seqNo = (Long) args[6];
+
+                    Result result = null;
+                    for (Result r : Result.values()) {
+                        if (r.getLowercase().equals(args[4])) {
+                            result = r;
+                            break;
+                        }
+                    }
+
+                    UpdateResponse updateResponse = null;
+                    if (shardInfo != null && seqNo != null) {
+                        updateResponse = new UpdateResponse(shardInfo, shardId, type, id, seqNo, version, result);
+                    } else {
+                        updateResponse = new UpdateResponse(shardId, type, id, version, result);
+                    }
+
+                    GetResult get = (GetResult) args[7];
+                    if (get != null) {
+                        GetResult getResult = new GetResult(index, type, id, version, get.isExists(), get.sourceRef(), get.getFields());
+                        updateResponse.setGetResult(getResult);
+                    }
+                    return updateResponse;
+                });
+
+        DocWriteResponse.declareParserFields(PARSER);
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> {
+            XContent xContent = p.contentType().xContent();
+            try (XContentBuilder builder = XContentBuilder.builder(xContent)) {
+                // "get" field contains an embedded version of {@link GetResult} and requires
+                // to be parsed as if it was a full XContent object.
+                XContentHelper.copyCurrentStructure(builder.generator(), p);
```

Comment:
> because the xContentType can differ between the response and the get section?

No, this has been a crappy hack around the fact that `GetResult` can be rendered as a `ToXContentObject` using `toXContent()` or as a fragment using `toXContentEmbedded()`.

I changed that so that `GetResult` now has a `fromXContentEmbedded()` parsing method that can be used in UpdateResponse's object parser.

Note that `GetResult.toXContent()` delegates all the work to `fromXContentEmbedded()`. Please let me know what you think about this.